flips/20210916-nft-metadata.md

```diff
@@ -0,0 +1,354 @@
+sellable items. With all the oncoming marketplaces/auction houses and things like that 
+coming to Flow a good standard for NFTs will help everybody. 
+
+## User Benefit
+We feel that a standard that authors can implement that enables easy consumption of metadata 
+will lead to more applications that integrate NFTs. This will make NFTs intrinsically more 
+useful and valuable to the end users holding the NFTs.
+
+A user will be much more likely to transfer their NFT to other marketplaces and sell their NFT 
+in different ways with this solution. Without it all marketplaces need to hardcode support 
+for all NFTs directly. 
+
+It will also open up the possibility to standardize on some shared structs like Royalty 
+and SaleItem/Auction.
+
+## Design Proposal
+
+The core of this proposal is to add the following interface and have the NFT resource implement it
+```
+pub resource interface ViewResolver {
+  pub fun getViews() : [Type]
+  pub fun resolveView(_ view:Type): AnyStruct
+}
+```
+
+We will also add default implementations for these methods so that no compatibility will be broken 
+for existing NFT solutions. The new NFT interface could look something like this: 
+```
+pub resource NFT: INFT, ViewResolver {
+    pub let id: UInt64
+    pub fun getViews(): [Type] {
+       return []
+    }
+    pub fun resolveView(_ type: Type): AnyStruct {
+        return nil
+    }
+}
+```
+
+We propose that all metadata be expressed in Cadence types. The `pub fun getViews(): [Type]` 
+method allows for discovery of the various types of metadata supported by the NFT by returning a `Type` 
+for each supported type of metadata. An instance of any supported type could then be retrieved by 
+calling the `pub fun resolveView(_ type: Type): AnyStruct` method passing the desired type in as 
+an argument.
+
+This proposal by no means precludes the use of member variable on the NFT resource itself. In fact,
+ideally all immutable data for an NFT is stored as immutable members of the NFT resource and views 
+are resolved and created by accessing those immutable members.
+
+### Supporting other resources besides NFTs
+In order to support any other resource and resource collections, like marketplaces and fungible tokens, the following new interface is proposed:
+```
+pub resource interface ViewResolverCollection {
+    pub fun borrowViewResolver(id: UInt64): &{ViewResolver}
+    pub fun getIDs(): [UInt64]
+}
+```
+
+This will allow any collection to be exposed as view resolvable.
+
+For single resources that are stored as is like a fungible token or a profile contract you can just implement the `ViewResolver` interface directly.
+
+### Converting between different views
+
+Standards will evolve and new views will be added. In order to make it easier for maintainers we propose a mechanism to register a converter from one View to another:
+
+```
+pub struct interface ViewConverter {
```

Comment:
Not sure if this is very useful. You wouldn't be able to convert _anything_ to _anything_ else (ie: converting an `Animal` to a `IPFSImage`), and in the case that you're trying to convert `A` to `B` you could instead just look for a `ViewResolver` that supports `B` and ask for the `B` directly rather than finding a `ViewResolver` for `A`, then iterating through `ViewConverter`s until you found one that converts `A` to `B` and then converting it.

Reply:
If you register a function that knows how to convert an Animal to an IPFSImage then you would be able to.

The use case here is if you have an NFT that knows how to resolve A and then a function from A -> B. So then you can also resolve that NFT as B via the ViewConverter.

Reply:
But why not just resolve the `B` view and cut out the conversion?

Reply:
WIth this in place, when I want to get a view of type `B` i have to first check the view resolver methods to see if there is a resolver for `B`, and if not - I then have to cycle through all of the `ViewConverter`s to find one that converts to `B` from another type that I also know. It just seems like a lot of steps vs. just being able to resolve the `B` view.

Reply:
I think here also can use `ViewResolver` maybe, instead of `to` and `convert`. 

If I have `ArtistConverter` instead of `convert` function, it can just have a `ViewResolver` that can provide destination formats.

Reply:
> WIth this in place, when I want to get a view of type B i have to first check the view resolver methods to see if there is a resolver for B, and if not - I then have to cycle through all of the ViewConverters to find one that converts to B from another type that I also know. It just seems like a lot of steps vs. just being able to resolve the B view.

if you want B, I think you will have a list of converters that converting to B, you will get a set of `from`s, then check if NFT resolves any of them I guess.

Reply:
I think this is more like: Imagine TopShot not supporting Blocto's `[BloctoDisplay]` view , so blocto can write a converter, use TopShot view then converter to get `[BloctoDisplay]`